perfkitbenchmarker/flag_util.py

```diff
@@ -0,0 +1,198 @@
+      return self
+
+    def next(self):
+      if self.idx < self.integer_list.length:
+        val = self.integer_list[self.idx]
+        self.idx += 1
+        return val
+      else:
+        raise StopIteration()
+
+
+class IntegerListParser(flags.ArgumentParser):
+  """Parse a string containing a comma-separated list of nonnegative integers.
+
+  The list may contain single integers and dash-separated ranges. For
+  example, "1,3,5-7" parses to [1,3,5,6,7].
+
+  Can pass the flag on_nonincreasing to the constructor to tell it
+  what to do if the list is nonincreasing. Options are
+    - None: do nothing.
+    - "Warn": log a warning.
+    - "Exception": raise a ValueError.
+
+  As a special case, instead of a string, can pass a list of integers
+  or an IntegerList. In these cases, the return value iterates over
+  the same integers as were in the argument.
+  """
+
+  syntactic_help = ('A comma-separated list of nonnegative integers or integer '
+                    'ranges. Ex: 1,3,5-7 is read as 1,3,5,6,7.')
+
+  def __init__(self, on_nonincreasing=None):
+    super(IntegerListParser, self).__init__()
+
+    self.on_nonincreasing = on_nonincreasing
+
+  def Parse(self, inp):
+    """Parse an integer list.
+
+    Args:
+      inp: a string, a list, or an IntegerList.
+
+    Returns:
+      An iterable of integers.
+
+    Raises:
+      ValueError if inp doesn't follow a format it recognizes.
+    """
+
+    if isinstance(inp, IntegerList):
+      return inp
+    elif isinstance(inp, list):
+      return IntegerList(inp)
+
+    groups = inp.split(',')
+    result = []
+
+    for group in groups:
+      match = INTEGER_GROUP_REGEXP.match(group)
+      if match is None:
+        raise ValueError('Invalid integer list %s', inp)
+      elif match.group(2) is None:
+        val = int(match.group(1))
+
+        if (self.on_nonincreasing and
+            len(result) > 0 and
+            val <= result[-1]):
+          if self.on_nonincreasing == "Warn":
+            logging.warning('Integer list %s is not increasing', inp)
+          elif self.on_nonincreasing == "Exception":
+            raise ValueError('Integer list %s is not increasing', inp)
+
+        result.append(val)
+      else:
+        low = int(match.group(1))
+        high = int(match.group(3))
+
+        if self.on_nonincreasing:
+          if high <= low or (len(result) > 0 and low <= result[-1]):
+            if self.on_nonincreasing == "Warn":
+              logging.warning('Integer list %s is not increasing', inp)
+            elif self.on_nonincreasing == "Exception":
+              raise ValueError('Integer list %s is not increasing', inp)
+
+        result.append((low, high))
+
+    return IntegerList(result)
+
+  def Type(self):
+    return 'integer list'
+
+
+class IntegerListSerializer(flags.ArgumentSerializer):
+  def Serialize(self, il):
+    return ','.join([str(val) if isinstance(val, int) or isinstance(val, long)
+                     else '%s-%s' % (val[0], val[1])
```

Comment:
Consider introducing namedtuple subclasses that support serialization to string and iteration to represent the two cases (int or range) that can be stored in the list. Probably overkill.